src/controller/controller.cpp

```diff
@@ -213,51 +185,32 @@ void Controller::runCommandHandler(const std::vector<ElectronicID::ptr>& availab
 
 void Controller::onCertificatesLoaded()
 {
-    CardEventMonitorThread* cardEventMonitorThread =
-        new CardEventMonitorThread(this, std::string(commandType()));
-    saveChildThreadPtrAndConnectFailureFinish(cardEventMonitorThread);
-    cardEventMonitorThreadKey = uintptr_t(cardEventMonitorThread);
+    auto* cardEventMonitorThread = new CardEventMonitorThread(this, commandType());
+    connect(cardEventMonitorThread, &ControllerChildThread::failure, this,
+            &Controller::onCriticalFailure);
+    connect(this, &Controller::stopCardEventMonitorThread, cardEventMonitorThread,
+            &WaitForCardThread::requestInterruption);
     connect(cardEventMonitorThread, &CardEventMonitorThread::cardEvent, this, &Controller::onRetry);
     cardEventMonitorThread->start();
 }
 
-void Controller::stopCardEventMonitorThread()
-{
-    if (cardEventMonitorThreadKey) {
-        try {
-            auto cardEventMonitorThread = childThreads.at(cardEventMonitorThreadKey);
-            cardEventMonitorThreadKey = 0;
-            if (cardEventMonitorThread) {
-                interruptThread(cardEventMonitorThread);
-            }
-        } catch (const std::out_of_range&) {
-            qWarning() << "Card event monitor thread" << cardEventMonitorThreadKey
-                       << "is missing from childThreads map in stopCardEventMonitorThread()";
-            cardEventMonitorThreadKey = 0;
-        }
-    }
-}
-
 void Controller::disposeUI()
 {
-    if (window) {
-        window->disconnect();
-        // As the Qt::WA_DeleteOnClose flag is set, the dialog is deleted automatically.
-        window->close();
-        window = nullptr;
-    }
+    delete window;
```

Comment:
This simplification reduces safety, so I suggest to revert this. If you really want immediate deletion, you can use `window->deleteLater()` after `close()`:

```c++
    if (window) {
        window->disconnect();
        // As the Qt::WA_DeleteOnClose flag is set, this will delete the dialog automatically.
        window->close();
        window->deleteLater();
        window = nullptr;
    }
```

Reply:
It is perfectly valid free and delete null pointer.
https://en.cppreference.com/w/cpp/memory/new/operator_delete.html

> 1) If ptr is null, does nothing. Otherwise, reclaims the storage allocated by the earlier call to [operator new](https://en.cppreference.com/w/cpp/memory/new/operator_new.html).

Reply:
Right, but when using `close()` and  `deleteLater()` it is required. And Qt docs warn that not using `deleteLater()` may crash if the object is deleted while it is handling an event.